pkg/kv/kvserver/replica_circuit_breaker.go

```diff
@@ -0,0 +1,183 @@
+	defaultReplicaCircuitBreakerSlowReplicationThreshold,
+	func(d time.Duration) error {
+		if d < 0 {
+			return errors.New("duration must be non-negative")
+		}
+		return nil
+	},
+)
+
+type replicaCircuitBreaker struct {
+	ambCtx  log.AmbientContext
+	stopper *stop.Stopper
+	r       replicaInCircuitBreaker
+	st      *cluster.Settings
+	wrapped *circuit.Breaker
+}
+
+func (br *replicaCircuitBreaker) enabled() bool {
+	return replicaCircuitBreakerSlowReplicationThreshold.Get(&br.st.SV) > 0 &&
+		br.st.Version.IsActive(context.Background(), clusterversion.ProbeRequest)
+}
+
+func (br *replicaCircuitBreaker) newError() error {
+	return br.r.rangeUnavailableError()
+}
+
+func (br *replicaCircuitBreaker) TripAsync() {
+	if !br.enabled() {
+		return
+	}
+
+	_ = br.stopper.RunAsyncTask(
+		br.ambCtx.AnnotateCtx(context.Background()), "trip-breaker",
+		func(ctx context.Context) {
+			br.wrapped.Report(br.newError())
+		},
+	)
+}
+
+type neverTripSignaller struct{}
+
+func (s neverTripSignaller) Err() error         { return nil }
+func (s neverTripSignaller) C() <-chan struct{} { return nil }
+
+func (br *replicaCircuitBreaker) Signal() interface {
+	Err() error
+	C() <-chan struct{}
+} {
+	if !br.enabled() {
+		return neverTripSignaller{}
+	}
+	return br.wrapped.Signal()
+}
+
+func newReplicaCircuitBreaker(
+	cs *cluster.Settings,
+	stopper *stop.Stopper,
+	ambientCtx log.AmbientContext,
+	r replicaInCircuitBreaker,
+) *replicaCircuitBreaker {
+	br := &replicaCircuitBreaker{
+		stopper: stopper,
+		ambCtx:  ambientCtx,
+		r:       r,
+		st:      cs,
+	}
+
+	br.wrapped = circuit.NewBreaker(circuit.Options{
+		Name:       redact.Sprintf("breaker"), // log bridge has ctx tags
+		AsyncProbe: br.asyncProbe,
+		EventHandler: &circuit.EventLogger{
+			Log: func(buf redact.StringBuilder) {
+				log.Infof(ambientCtx.AnnotateCtx(context.Background()), "%s", buf)
+			},
+		},
+	})
+
+	return br
+}
+
+type probeKey struct{}
+
+func isCircuitBreakerProbe(ctx context.Context) bool {
+	return ctx.Value(probeKey{}) != nil
+}
+
+func withCircuitBreakerProbeMarker(ctx context.Context) context.Context {
+	return context.WithValue(ctx, probeKey{}, probeKey{})
+}
+
+func (br *replicaCircuitBreaker) asyncProbe(report func(error), done func()) {
+	bgCtx := br.ambCtx.AnnotateCtx(context.Background())
+	if err := br.stopper.RunAsyncTask(
+		bgCtx,
+		"replica-probe",
+		func(ctx context.Context) {
+			defer done()
+
+			if !br.enabled() {
+				report(nil)
+				return
+			}
+
+			err := sendProbe(ctx, br.r)
+			report(err)
+		}); err != nil {
+		done()
+	}
+}
+
+func sendProbe(ctx context.Context, r replicaInCircuitBreaker) error {
+	ctx = withCircuitBreakerProbeMarker(ctx)
+	desc := r.Desc()
+	if !desc.IsInitialized() {
+		return nil
+	}
+	ba := roachpb.BatchRequest{}
+	ba.Timestamp = r.Clock().Now()
+	ba.RangeID = r.Desc().RangeID
+	probeReq := &roachpb.ProbeRequest{}
+	probeReq.Key = desc.StartKey.AsRawKey()
+	ba.Add(probeReq)
+	thresh, ok := r.slowReplicationThreshold(&ba)
```

Comment:
nit: consider doing this at the start of the function, since it negates the entire function.